File: leetcode/Edit Distance.cpp
#include <iostream>
#include <vector>
using namespace std;
int min(int a,int b,int c){
	return a<(b<c?b:c)?a:(b<c?b:c);
}
int minDistance(string word1, string word2) {
	int l1=word1.length()+1;
	int l2=word2.length()+1;
	
	vector<vector<int> > f(l1,vector<int>(l2));
	for(int i=0;i<l1;i++) f[i][0]=i;
	for(int i=0;i<l2;i++) f[0][i]=i;
	
	for(int i=1;i<l1;i++)
	{
		for(int j=1;j<l2;j++)
		{
			if(word1[i]==word2[j]) f[i][j]=f[i-1][j-1];
			else
				f[i][j]=f[i-1][j-1]+1;
			f[i][j]=min(f[i][j],f[i-1][j]+1,f[i][j-1]+1);
		}
	} 
	return f[l1-1][l2-1]; 
}

int main(int argc, char *argv[]) {
	string w1="asdf";
	string w2="asdfasdf";
	cout<<minDistance(w1,w2);
	//cout<<min(3,2,1);
}

File: leetcode/leetcode/src/com/buptaaa/term1/EditDistance.java
package com.buptaaa.term1;

/**
 * Created by shiyanghuang on 15/6/21.
 */
public class EditDistance {
    public int EditDistance(String word1, String word2) {
        int l1 = word1.length();
        int l2 = word2.length();
        if(word1.isEmpty())
            return l2;
        if(word2.isEmpty())
            return l1;
        int[][] dp = new int[l1+1][l2+1];
        //Base condition
        for(int i =0; i<=l1; i++)
            dp[i][0] = i;
        for(int j=1; j<=l2; j++)
            dp[0][j] = j;
        for(int i=1; i<=l1 ; i++)
        {
            for(int j=1; j<=l2; j++)
            {
                if(word1.charAt(i-1) == word2.charAt(j-1))
                    dp[i][j] = dp[i-1][j-1];
                else
                    dp[i][j] = Math.min( Math.min(dp[i-1][j] + 1, dp[i][j-1] + 1), dp[i-1][j-1] + 1);
            }
        }
        System.out.print("    ");
        for (int i = 0; i < l2; i++) {
            System.out.print(word2.charAt(i) + " ");
        }
        System.out.println();
        for (int i = 0; i <= l1; i++) {
            if (i >= 1) {
                System.out.print(word1.charAt(i - 1) + " ");
            } else {
                System.out.print("  ");
            }

            for (int j = 0; j <= l2; j++) {
                System.out.print(dp[i][j] + " ");
            }
            System.out.println();
        }
        return dp[l1][l2];
    }

    public static void main(String[] args) {
        System.out.println(new EditDistance().EditDistance("asdf", "qwddfrr"));
    }
}
